common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -289,15 +289,18 @@ public UTF8String substringSQL(int pos, int length) {
    * Returns whether this contains `substring` or not.
    */
   public boolean contains(final UTF8String substring) {
-    if (substring.numBytes == 0) {
+    final int len = substring.numBytes;
+    if (len == 0) {
       return true;
     }
 
-    byte first = substring.getByte(0);
-    for (int i = 0; i <= numBytes - substring.numBytes; i++) {
-      if (getByte(i) == first && matchAt(substring, i)) {
-        return true;
-      }
+    final byte first = substring.getByte(0);
+    final Object base = this.base;
+    long offset = this.offset;
+    final long end = offset + numBytes - len;
+    for (; offset <= end; offset++) {
+      if (Platform.getByte(base, offset) == first && ByteArrayMethods.arrayEquals(
```

Comment:
I was just wondering if finding the first byte in this UTF8String which is equal to first and then loop it from there will be helpful or unnecessary overhead.

Reply:
You mean remove the first check above in the && above? Yes tried that and is substantially slower because arrayEquals does much more like aligned reads etc (and perhaps also does not get inlined).

Reply:
No no. I am saying the reverse that is do keep it and introduce the basefirst byte also in comparison. I mean find out the first till it is equal to first of the argument substring and then do ByteArrayMethods.arrayEquals with the new offset of the base string. Looks like for long string where the first byte is not present till very end of the string then it will be a cheaer contains call.

Reply:
> I mean find out the first till it is equal to first of the argument substring

Didn't get this. The "Platform.getByte(base, offset) == first" is doing that already. Can you write sample code to show what you are saying?

Reply:
Yes. Got confused. Ignore.